internal/btf/btf.go

```diff
@@ -193,67 +194,79 @@ func loadSpecFromELF(file *internal.SafeELFFile) (*Spec, error) {
 // transforms them to be indexed by function. Retrieves function names from
 // the BTF spec.
 func (spec *Spec) splitExtInfos(info *extInfo) error {
-
 	ofi := make(map[string]FuncInfo)
 	oli := make(map[string]LineInfos)
 	ocr := make(map[string]CoreRelos)
 
 	for secName, secFuncs := range info.funcInfos {
 		// Collect functions from each section and organize them by name.
+		var funcs []*Func
 		for _, fi := range secFuncs {
-			name, err := fi.Name(spec)
+			typ, err := spec.TypeByID(fi.TypeID)
 			if err != nil {
-				return fmt.Errorf("looking up function name: %w", err)
+				return err
 			}
 
-			// FuncInfo offsets are scoped to the ELF section. Zero them out
-			// since they are meaningless outside of that context. The linker
-			// will determine the offset of the function within the final
-			// instruction stream before handing it off to the kernel.
-			fi.InsnOff = 0
+			fn, ok := typ.(*Func)
+			if !ok {
+				return fmt.Errorf("type ID %d is a %T, but expected a Func", fi.TypeID, typ)
+			}
 
-			ofi[name] = fi
+			// C doesn't have anonymous functions, but check just in case.
+			if fn.Name == "" {
+				return fmt.Errorf("func with type ID %d doesn't have a name", fi.TypeID)
+			}
+
+			ofi[fn.Name] = FuncInfo{fn}
+			funcs = append(funcs, fn)
+		}
+
+		// Consider an ELF section that contains 3 functions (a, b, c)
+		// at offsets 0, 10 and 15 respectively. Offset 5 will return function a,
+		// offset 12 will return b, offset >= 15 will return c, etc.
+		sort.Slice(secFuncs, func(i, j int) bool {
```

Comment:
Could we get away without this? They appear in the list here in the order they were read from .BTF.ext, which should already be ordered by offset. I'm fine with leaving this in, we're at the mercy of the compiler otherwise.

Reply:
I went back and forth on this as well. Sorting here has two purposes: it "documents" that funcForInstructions relies on this ordering to work correctly and at the same time enforces that invariant. That seems worth it to me.

Reply:
Yep, makes sense. :+1: